text/0696-ember-cli-github-actions.md

```diff
@@ -0,0 +1,58 @@
+---
+Stage: Accepted
+Start Date: 2021-01-04
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember CLI
+RFC PR: https://github.com/emberjs/rfcs/pull/696
+---
+
+# Replace Travis CI with GitHub Actions in generated Ember CLI projects
+
+## Summary
+
+Travis CI is no longer an easy setup for our users.
+GitHub Actions is the obvious replacement.
+
+## Motivation
+
+With the change of ownership, Travis CI is cracking down on open source usage.
+https://travis-ci.org has limited the number of builds for open source, which means in practice, most builds are queued for hours.
+https://travis-ci.org is also slated for shutdown soon.
+https://www.travis-ci.com for open source has been converted to free trial.
+After the initial credits are used, you must submit justification for more.
+There have been reports that they are refusing requests for more open source credits https://twitter.com/james_hilliard/status/1336081776691843072.
+
+## Detailed design
+
+We will replace the .travis.yml with a .github/workflows/ci.yml.
+It will be configured as close as possible to the existing Travis CI setup.
+Unlike Travis CI, there is no enable or authorization step.
+The presence of a file in .github/workflows/ is automatically recognized and run if hosted on GitHub.
+
+## How we teach this
+
+A blog post could be sufficient explaining this switch.
+Existing projects can continue to use Travis CI in its limited capacity.
+Those using ember-cli-update could choose to take the new GitHub Actions file or continue using the .travis.yml.
+I don't think we cover CI or Travis CI in our docs.
+
+## Drawbacks
+
+Some CI systems are agnostic to code hosting location, but I don't think you can use GitHub Actions without also placing the code there.
+This further ingrains the GitHub platform to people that may prefer a different code hosting provider.
+
+GitHub Actions also has a limit how much an open source project can use, but it resets every month.
+I haven't heard of anyone hitting the limit though.
+
+GitHub Actions doesn't support retrying individual jobs like Travis CI.
+This could impact some user's monthly budget.
+
+## Alternatives
+
+CircleCI and GitLab are other popular alternatives.
+Perhaps one would be better as a recommendation from us?
+
+We could also refrain from suggesting/including any CI config files.
```

Comment:
This would mean removing `.travis.yml` from default blueprints without a replacement, wouldn't it? This would be my preferred short-term solution. Additionally we should experiment with alternative solutions to provide an easy setup of CI pipelines. But no need to do this kind of experiments in Ember CLI itself.

Reply:
The reality is that the majority of Ember addons are living on GitHub, and that they have been using TravisCI by default so far. Not including a config means that new addons don't get CI by default, and existing addons might have their CI config removed. That would not be a good solution IMHO.

I agree with Kelly that the MVP here is replacing the TravisCI config with a default GitHub Action config, and if that isn't used because of using some other hosting provider then people can delete it.

This does not mean we should stop considering other options, like the generator that was mentioned. But we need a short-term solution now, and this RFC has a much smaller scope, which makes it much more likely that we can implement it soonish.

Reply:
> The reality is that the majority of Ember addons are living on GitHub, and that they have been using TravisCI by default so far.

I agree that this is true for open source addons. But I'm not sure if it is still true if also considering internal addons. We are making heavy use of private addons in the project I'm currently working on. Not using GitHub or TravisCI for it. Not sure how it is for others.

The argument also doesn't take applications into account. CI configuration is currently not only included in addon but also in application blueprints. We are not using GitHub but GitLab at work. I _guess_ many other teams do as well.

> Not including a config means that new addons don't get CI by default

Long time since I created a new addon, which was using TravisCI. But if I recall correctly maintainer needed to enable the TravisCI for the new project explicitly. I don't see a big difference between doing that manual step or running a setup script like `yarn create-github-actions-for-ember-addon`.

> and existing addons might have their CI config removed.

The motivation section of this RFC states that the existing CI configuration is unusable. So deleting the `.travis.yml` when updating Ember CLI blueprints should not make it worth. Additionally nothing prevents a maintainer from _not_ following that blueprint change.

> I agree with Kelly that the MVP here is replacing the TravisCI config with a default GitHub Action config, and if that isn't used because of using some other hosting provider then people can delete it.

The file will be recreated by `ember-cli-update` everytime it changes in default blueprints. For CI configuration this is atleast once per LTS version. It's easy to miss on upgrades. It quickly gets annoying. We stopped deleting the existing `.travis.yml` file at work therefore. @mehulkar raised the same concern in https://github.com/emberjs/rfcs/issues/596#issuecomment-632909864.

> This does not mean we should stop considering other options, like the generator that was mentioned. But we need a short-term solution now, and this RFC has a much smaller scope, which makes it much more likely that we can implement it soonish.

As far as I understand we agree that `.travis.yml` should be removed from default blueprints. We seems to disagree if a GitHub Actions workflow should be added as a replacement or not. Taking scope into account not proving a replacement as part of the blueprint is smaller than providing one.

Reply:
> Taking scope into account not proving a replacement as part of the blueprint is smaller than providing one.

it not smaller at all, because "should we include a CI config at all?" is a much larger discussion than "should we include a CI config for the most used CI service?"

I'm aware that for internal apps and addons different rules apply, but I am primarily worried about public addons not running CI anymore because of the change you propose. If internal apps and addons break because they are not running CI anymore then that's an internal problem, but if public addons stop working then that becomes a problem of everyone.